controllers/core/domain_controller.go

```diff
@@ -0,0 +1,163 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	openfunction "github.com/openfunction/apis/core/v1alpha2"
+	"github.com/openfunction/pkg/util"
+)
+
+const (
+	reloadTimestamp = "reloadtimestamp"
+)
+
+// DomainReconciler reconciles a Domain object
+type DomainReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+	ctx    context.Context
+}
+
+func NewDomainReconciler(mgr manager.Manager) *DomainReconciler {
+
+	r := &DomainReconciler{
+		Client: mgr.GetClient(),
+		Scheme: mgr.GetScheme(),
+		Log:    ctrl.Log.WithName("controllers").WithName("Domain"),
+	}
+
+	return r
+}
+
+//+kubebuilder:rbac:groups=core.openfunction.io,resources=domains,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core.openfunction.io,resources=domains/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Serving object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *DomainReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.ctx = ctx
+	log := r.Log.WithValues("Domain", req.NamespacedName)
+
+	var d openfunction.Domain
+
+	if err := r.Get(ctx, req.NamespacedName, &d); err != nil {
+		if util.IsNotFound(err) {
+			log.V(1).Info("Domain deleted")
+		}
+		return ctrl.Result{}, util.IgnoreNotFound(err)
+	}
+
+	svc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      d.Name,
+			Namespace: d.Namespace,
+		},
+	}
+
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(svc), svc); util.IgnoreNotFound(err) != nil {
+		log.Error(err, "Failed to get service")
+		return ctrl.Result{}, err
+	}
+
+	op, err := controllerutil.CreateOrUpdate(context.TODO(), r.Client, svc, r.svcMutate(&d, svc))
+	if err != nil {
+		log.Error(err, "Failed to CreateOrUpdate domain")
+		return ctrl.Result{}, err
+	}
+
+	if err := r.updateFunction(); err != nil {
+		log.Error(err, "Failed to update function")
+		return ctrl.Result{}, err
+	}
+
+	log.V(1).Info(fmt.Sprintf("Domain %s", op))
+	return ctrl.Result{}, nil
+}
+
+func (r *DomainReconciler) svcMutate(d *openfunction.Domain, svc *corev1.Service) controllerutil.MutateFn {
+
+	return func() error {
+
+		svc.Spec.ExternalName =
+			fmt.Sprintf("%s.%s.svc.cluster.local", d.Spec.Ingress.Service.Name, d.Spec.Ingress.Service.Namespace)
```

Comment:
`.cluster.local` can be removed because users might customize his own

Reply:
The `externalName` of ExternalName service need the `.cluster.local`.

Later I will consider how to solve this case.